File: PrimaMateria/PrimaMateria/Classes/XTRMainViewController.m
//
//  XTRMainViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright (c) 2011 xTrensa. All rights reserved.
//

#import "PrimaMateria.h"

@interface XTRMainViewController ()
- (void)customizeTabBarItems;
- (void)showSplash;
@end

@implementation XTRMainViewController

- (void)customizeTabBarItems {
    CGRect rect = self.tabBar.frame;
    CGSize newSize = CGSizeMake(90, rect.size.height);
    UIImage *selectionImage = [UIImage imageWithColor: UIColor.lightGrayColor size: newSize];
    UIImageView *imageView = [[UIImageView alloc] initWithImage: selectionImage];
    
    UIGraphicsBeginImageContextWithOptions(imageView.bounds.size, NO, [UIScreen mainScreen].scale);
    
    [[UIBezierPath bezierPathWithRoundedRect:imageView.bounds cornerRadius: 8.0] addClip];
    [selectionImage drawInRect:imageView.bounds];
    
    self.tabBar.selectionIndicatorImage = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
}

-(void) showSplash {
    BOOL defaultState = [XTRPropertiesStore retreiveSplashScreenState];
    
    if (defaultState) {
        UIStoryboard *storyboard =[UIStoryboard storyboardWithName:MAIN_STORY_BOARD bundle:nil];
        XTRSplashViewController *splashViewController = [storyboard instantiateViewControllerWithIdentifier:NSStringFromClass([XTRSplashViewController class])];
        [self.view addSubview : splashViewController.view];
    }
}

#pragma mark - UITabBarControllerDelegate Methods

- (BOOL)tabBarController:(UITabBarController *)tabBarController shouldSelectViewController:(UIViewController *)viewController {

    BOOL defaultState = [XTRPropertiesStore retreiveShowTransitionsState];
    
    if (defaultState) {
        UIViewController *currentVC = tabBarController.selectedViewController;
        
        if (currentVC == viewController) {
            return NO;
        }
        
        [UIView beginAnimations:@"View Flip" context:nil];
        [UIView setAnimationDuration:1.5];
        [UIView setAnimationCurve:UIViewAnimationCurveEaseInOut];
        [UIView setAnimationTransition: UIViewAnimationTransitionCurlUp forView:tabBarController.view cache:YES];
        [currentVC viewWillAppear:YES];
        [viewController viewWillDisappear:YES];
        [viewController viewDidDisappear:YES];
        [currentVC viewDidAppear:YES];
        [UIView commitAnimations];
    }
	return YES;
}

#pragma mark - View Management Methods

- (void)viewDidLoad {
    [super viewDidLoad];
    [self showSplash];
    [self customizeTabBarItems];

    //self.delegate = self;
}

- (BOOL) shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

#pragma mark - Memory Management Methods

@end

File: PrimaMateria/PrimaMateria/Classes/XTRMainViewController.swift
//
//  XTRMainViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/15/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

class XTRMainViewController : UITabBarController, UITabBarControllerDelegate {
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    func customizeTabBarItems() {
        let rect = tabBar.frame
        let newSize = CGSize(width: 120, height: rect.size.height)
        let selectionImage = UIImage.imageFromColor(UIColor.lightGray, andSize: newSize)
        let imageView = UIImageView(image: selectionImage)
        
        UIGraphicsBeginImageContextWithOptions(imageView.bounds.size, false, UIScreen.main.scale)
        
        UIBezierPath(roundedRect: imageView.bounds, cornerRadius: 8.0).addClip()
        
        selectionImage.draw(in: imageView.bounds)
        
        tabBar.selectionIndicatorImage = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
    }
    
    func showSplash() {
        let defaultState = XTRPropertiesStore.splashScreenState
        
        if defaultState {
            let splashViewController : XTRSplashViewController = XTRAppDelegate.storyboard().instantiateViewController(withIdentifier: XTRSplashViewController.nameOfClass) as! XTRSplashViewController
            view.addSubview(splashViewController.view)
        }
    }
    
    // MARK: - Action Methods
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        showSplash()
        if #available(iOS 11.0, *) {
            navigationController?.navigationBar.prefersLargeTitles = true
        } else {
            customizeTabBarItems()
        }
        
        delegate = self
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
}

extension XTRMainViewController { // UITabBarControllerDelegate Methods
    
    func tabBarController(_ tabBarController: UITabBarController, animationControllerForTransitionFrom fromVC: UIViewController, to toVC: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        let defaultState = XTRPropertiesStore.showTransitionsState
        
        if defaultState {
            //return XTRPopoutAnimationController()
            // XTRSlideDownAnimationController()
            return XTRSimpleTransitionController()
        }
        return nil
    }

}
